File: Wells/Source/VievModels/Entities/WellsViewModel.vb
Imports System.Reflection
Imports Wells.YPFModel
Imports Wells.ViewBase
Imports NPOI.XSSF.UserModel
Imports System.IO
Imports System.Xml.Serialization

Public Class WellsViewModel
    Inherits EntitiesViewModel(Of Well)

    Private _SelectedEntity As Well
    Private _SelectedEntities As IEnumerable(Of Well)
    Private _PremadeGraphics As List(Of PremadeSeriesInfoCollection)

    Sub New()
        MyBase.New(Nothing)
        IsNewCommandEnabled = True
        IsRemoveCommandEnabled = True
        FilterCollection = New FilterCollection(Of Well)
        Initialize()
        _Entities = _Repository.Wells.All
        _ShowWellPanel = True
        _PremadeGraphics = ReadPremadeGraphics()
    End Sub

    Protected Overrides Sub OnSetView(view As IView)
        MyBase.OnSetView(view)
        AddHandler _Control.MainWindow.PremadeGraphicsChanged, AddressOf OnPremadeGraphicsChanged
    End Sub

    Public Overrides ReadOnly Property FilterProperties As Dictionary(Of String, PropertyInfo)
        Get
            Return Well.Properties
        End Get
    End Property

    Public Overrides Property SelectedEntity As Well
        Get
            Return _SelectedEntity
        End Get
        Set
            SetValue(_SelectedEntity, Value)
            NotifyPropertyChanged(NameOf(WellExistsInfo))
        End Set
    End Property

    Public Overrides ReadOnly Property EditEntityCommand As ICommand = New RelayCommand(Sub()
                                                                                            Dim vm As New EditWellViewModel(SelectedEntity)
                                                                                            If _Control.MainWindow.OpenEditEntityDialog(vm) Then
                                                                                                UpdateEntites()
                                                                                            End If
                                                                                        End Sub, Function() SelectedEntity IsNot Nothing, AddressOf OnError)

    Public Overrides ReadOnly Property IsNewCommandEnabled As Boolean

    Public Overrides ReadOnly Property IsRemoveCommandEnabled As Boolean

    Public Overrides ReadOnly Property NewEntityCommand As ICommand = New RelayCommand(Sub()
                                                                                           Dim vm As New EditWellViewModel()
                                                                                           If _Control.MainWindow.OpenEditEntityDialog(vm) Then
                                                                                               UpdateEntites()
                                                                                           End If
                                                                                       End Sub, Function() IsNewCommandEnabled, AddressOf OnError)

    Public Overrides ReadOnly Property ImportEntitiesCommand As ICommand = New RelayCommand(Sub()
                                                                                                Dim wb As XSSFWorkbook = Nothing
                                                                                                Dim sheetIndex As Integer = -1

                                                                                                If OpenExcelFile(wb, sheetIndex) Then
                                                                                                    ReadExcelFile(wb, sheetIndex)
                                                                                                    UpdateEntites()

                                                                                                End If
                                                                                            End Sub, Function() IsNewCommandEnabled, AddressOf OnError, AddressOf CloseWaitingMessage)

    Public Overrides ReadOnly Property RemoveEntityCommand As ICommand = New RelayCommand(Sub()
                                                                                              If _Control.MainWindow.ShowMessageBox("¿Está seguro de eliminar este pozo?", "Eliminar") Then
                                                                                                  _Repository.Wells.Remove(SelectedEntity)
                                                                                                  UpdateEntites()
                                                                                              End If
                                                                                          End Sub, Function() SelectedEntity IsNot Nothing AndAlso IsRemoveCommandEnabled, AddressOf OnError)

    ReadOnly Property OpenPremadeGraphicCommand As ICommand = New RelayCommand(Sub(param)
                                                                                   Dim pg = CType(param, PremadeSeriesInfoCollection)
                                                                                   _Control.MainWindow.OpenGraphicsView(SelectedEntity, pg)
                                                                               End Sub, Function() True, AddressOf OnError)

    Private Sub UpdateEntites()
        _Entities = _Repository.Wells.All
        NotifyPropertyChanged(NameOf(Entities))
    End Sub

    Private Async Sub ReadExcelFile(workbook As XSSFWorkbook, sheetIndex As Integer)
        ShowWaitingMessage("Leyendo pozos del archivo Excel...")
        Dim wells = Await Task.Run(Function() ExcelReader.ReadWells(workbook, sheetIndex, _progress))
        CloseWaitingMessage()

        If wells.Any Then
            ShowWaitingMessage("Importando pozos...")
            Await Task.Run(Sub() _Repository.Wells.AddRangeAsync(wells))
            CloseWaitingMessage()

            ShowWaitingMessage("Guardando base de datos...")
            Await _Repository.SaveChangesAsync()

        End If

        workbook.Close()
        CloseWaitingMessage()
    End Sub

    Protected Overrides Sub CreateWellFilter()
        Dim wellFilter = New WellFilter(Of Well)(_Repository.Wells, True, WellType, WellProperty, SelectedWellName)
        OnCreatingFilter(wellFilter)
    End Sub

    Public Overrides Function GetContextMenu() As ContextMenu
        Dim menu = New ContextMenu()
        Dim editMenuItem As New MenuItem() With {.Header = "Editar...", .Command = EditEntityCommand}
        Dim removeMenuItem As New MenuItem() With {.Header = "Eliminar", .Command = RemoveEntityCommand}
        menu.Items.Add(editMenuItem)
        menu.Items.Add(New Separator)

        Dim graphicsMenuItem As New MenuItem() With {.Header = "Gráficos"}
        For Each pg In _PremadeGraphics
            Dim aMenuItem As New MenuItem() With {.Header = pg.Title, .Command = OpenPremadeGraphicCommand, .CommandParameter = pg}
            graphicsMenuItem.Items.Add(aMenuItem)
        Next
        menu.Items.Add(graphicsMenuItem)

        menu.Items.Add(New Separator)
        menu.Items.Add(removeMenuItem)
        Return menu
    End Function

    Private Function ReadPremadeGraphics() As List(Of PremadeSeriesInfoCollection)
        Dim filename = Path.Combine(Directory.GetCurrentDirectory, "PremadeGraphics.wpg")

        If File.Exists(filename) Then
            Dim serializer As New XmlSerializer(GetType(List(Of PremadeSeriesInfoCollection)))
            Dim entities As List(Of PremadeSeriesInfoCollection) = Nothing
            Using reader As New IO.StreamReader(filename)
                entities = CType(serializer.Deserialize(reader), List(Of PremadeSeriesInfoCollection))
            End Using

            Return entities
        End If
        Return Nothing
    End Function

    Private Sub OnPremadeGraphicsChanged()
        _PremadeGraphics = ReadPremadeGraphics()
        _Control.UpdateRowContextMenu()
    End Sub

    Overrides ReadOnly Property WellExistsInfo As String
        Get
            If _SelectedEntity IsNot Nothing Then
                Return If(_SelectedEntity.Exists, "Pozo existente", "Pozo inexistente")
            End If
            Return String.Empty
        End Get
    End Property
End Class

File: DesktopApp/src/ViewModels/Entities/WellsViewModel.cs
using NPOI.XSSF.UserModel;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using System.Windows.Controls;
using System.Windows.Input;
using System.Xml.Serialization;
using Wells.CoreView;
using Wells.CoreView.ViewInterfaces;
using Wells.View.Filters;
using Wells.View.Graphics;
using Wells.View.Importer;
using Wells.YPFModel;
using Wells.CorePersistence.Repositories;

namespace Wells.View.ViewModels
{
    public class WellsViewModel : EntitiesViewModel<Well>
    {
        private List<PremadeSeriesInfoCollection> _PremadeGraphics;

        public WellsViewModel() : base(null)
        {
            IsNewCommandEnabled = true;
            IsRemoveCommandEnabled = true;
            FilterCollection = new Filters.FilterCollection<Well>();
            Initialize();
            _Entities = Repository.Wells.All;
            _ShowWellPanel = true;
            _PremadeGraphics = CreatePremadeGraphicViewModel.ReadPremadeGraphics();
        }

        protected override void OnSetView(IView view)
        {
            base.OnSetView(view);
            MainWindow.PremadeGraphicsChanged += OnPremadeGraphicsChanged;
        } 

        public override string WellExistsInfo
        {
            get
            {
                if (_SelectedEntity != null)
                    return _SelectedEntity.Exists ? "Pozo existente" : "Pozo inexistente";
                return string.Empty;
            }
        }

        public override bool IsNewCommandEnabled { get; }

        public override bool IsRemoveCommandEnabled { get; }

        public override ICommand NewEntityCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var vm = new EditWellViewModel();
                    if (MainWindow.OpenEditEntityDialog(vm))
                        UpdateEntites();
                }, (obj) => IsNewCommandEnabled, OnError);
            }
        }

        public override ICommand EditEntityCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var vm = new EditWellViewModel(SelectedEntity);
                    if (MainWindow.OpenEditEntityDialog(vm))
                        UpdateEntites();

                }, (obj) => SelectedEntity != null, OnError);
            }
        }

        public override ICommand RemoveEntityCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if (MainWindow.ShowYesNoMessageBox("¿Está seguro de eliminar este pozo?", "Eliminar"))
                    {
                        Repository.Wells.Remove(SelectedEntity);
                        RepositoryWrapper.Instance.SaveChanges();
                        UpdateEntites();
                    }
                }, (obj) => SelectedEntity != null && IsRemoveCommandEnabled, OnError);
            }
        }

        public ICommand OpenPremadeGraphicCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    MainWindow.OpenGraphicsView(SelectedEntity, param as PremadeSeriesInfoCollection);
                }, (obj) => SelectedEntity != null, OnError);
            }
        }

        public ICommand OpenPiperShoellerGraphicCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                if (SelectedEntities != null && SelectedEntities.Any()) {
                    var vm = new PiperSchoellerGraphicViewModel(SelectedEntities);
                    MainWindow.OpenGraphicsView(vm); }
                else if (SelectedEntity != null) {
                    var vm = new PiperSchoellerGraphicViewModel(new List<Well>() { SelectedEntity });
                MainWindow.OpenGraphicsView(vm);
                                                                                         }
                }, (obj) => (SelectedEntities != null && SelectedEntities.Any()) || SelectedEntity != null, OnError);
            }
        }

        public override Dictionary<string, PropertyInfo> FilterProperties => Well.Properties;

        public override Well SelectedEntity { get => _SelectedEntity; set { SetValue(ref _SelectedEntity, value); NotifyPropertyChanged(nameof(WellExistsInfo)); } }

        protected override void SetCommandUpdates()
        {
            base.SetCommandUpdates();
            Add(nameof(SelectedEntity), OpenPremadeGraphicCommand );
            Add(nameof(SelectedEntity), OpenPiperShoellerGraphicCommand);
            Add(nameof(SelectedEntities), OpenPiperShoellerGraphicCommand);
        }

        public override ContextMenu GetContextMenu()
        {
            var menu = new ContextMenu();
            var editMenuItem = new MenuItem() { Header = "Editar...", Command = EditEntityCommand };
            var removeMenuItem = new MenuItem() { Header = "Eliminar", Command = RemoveEntityCommand };
            menu.Items.Add(editMenuItem);
            menu.Items.Add(new Separator());

            var graphicsMenuItem = new MenuItem() { Header = "Gráficos" };
            var piperMenuItem = new MenuItem() { Header = "Piper-Schöeller", Command = OpenPiperShoellerGraphicCommand, CommandParameter = SelectedEntities };
            graphicsMenuItem.Items.Add(piperMenuItem);
            foreach (var pg in _PremadeGraphics)
            {
                graphicsMenuItem.Items.Add(new Separator());
                var aMenuItem = new MenuItem() { Header = pg.Title, Command = OpenPremadeGraphicCommand, CommandParameter = pg };
                graphicsMenuItem.Items.Add(aMenuItem);
            }
            menu.Items.Add(graphicsMenuItem);

            menu.Items.Add(new Separator());
            menu.Items.Add(removeMenuItem);
            return menu;
        }

        protected override async Task ReadExcelFile(XSSFWorkbook workbook, int sheetIndex)
        {
            ShowWaitingMessage("Leyendo pozos del archivo Excel...");
            var wells = await Task.Run(() => ExcelReader.ReadWells(workbook, sheetIndex, null));
            CloseWaitingMessage();

            if (wells.Any())
            {
                ShowWaitingMessage("Importando pozos...");
                await Task.Run(() => Repository.Wells.AddRangeAsync(wells));
                CloseWaitingMessage();

                ShowWaitingMessage("Guardando base de datos...");
                await Repository.SaveChangesAsync();
            }

            workbook.Close();
            CloseWaitingMessage();
        }


        void OnPremadeGraphicsChanged(object sender, EventArgs args)
        {
            _PremadeGraphics = CreatePremadeGraphicViewModel.ReadPremadeGraphics();
            Control.UpdateRowContextMenu();
        }

        protected override void CreateWellFilter()
        {
            var wellFilter = new WellFilter<Well>(Repository.Wells, true, WellType, WellProperty, SelectedWellName);
            OnCreatingFilter(wellFilter);
        }

        protected override void UpdateEntites()
        {
            _Entities = Repository.Wells.All;
            NotifyPropertyChanged(nameof(Entities));
        }
    }
}
